### PR TITLE
Adding steps about how to fine tune on any custom dataset.

### PR DESCRIPTION
**Pull request by quic-swatia**

### CODE REVIEW

File: docs/source/finetune.md [hunk 1]
```diff
@@ -63,4 +63,38 @@ to visualise the data,
...
 tensorboard --logdir runs/<file> --bind_all
-```
\ No newline at end of file
+```
+
+## Fine-Tuning on custom dataset
+
+To run fine tuning for any user specific dataset, prepare the dataset using the following steps:
+
+    1) Create a  directory named 'dataset' inside efficient-transformers.
```
Comment (quic-meetkuma) [inline docs/source/finetune.md:72]: double space between "a" and "directory"

File: docs/source/finetune.md [hunk 2]
```diff
@@ -63,4 +63,38 @@ to visualise the data,
...
+```python
+def get_custom_dataset(dataset_config, tokenizer, split, context_length=None):
+
+    # load dataset
+    # based on split, retrieve only the specific portion of the dataset (train or eval) either here or at the last
+    
+    def apply_prompt_template():
+    
+    def tokenize():
+
```
Comment (quic-meetkuma) [inline docs/source/finetune.md:92]: Add some comment as "Implement tokenization and prepare inputs for the training."

File: docs/source/finetune.md [hunk 3]
```diff
@@ -63,4 +63,38 @@ to visualise the data,
...
+    8) While running fine tuning, pass argument "-–dataset custom_dataset" to finetune on custom dataset.   
+
+Template for get_custom_dataset() to be defined inside efficient-transformers/dataset/custom_dataset.py is as follows:
+
+```python
+def get_custom_dataset(dataset_config, tokenizer, split, context_length=None):
+
+    # load dataset
+    # based on split, retrieve only the specific portion of the dataset (train or eval) either here or at the last
+
```
Comment (quic-meetkuma) [inline docs/source/finetune.md:88]: Add one more comment as "Define a prompt template"

File: docs/source/finetune.md [hunk 4]
```diff
@@ -63,4 +63,38 @@ to visualise the data,
...
+Template for get_custom_dataset() to be defined inside efficient-transformers/dataset/custom_dataset.py is as follows:
+
+```python
+def get_custom_dataset(dataset_config, tokenizer, split, context_length=None):
+
+    # load dataset
+    # based on split, retrieve only the specific portion of the dataset (train or eval) either here or at the last
+    
+    def apply_prompt_template():
+
```
Comment (quic-meetkuma) [inline docs/source/finetune.md:90]: Add some comment as "Convert the raw input into format as per the template defined earlier."

File: docs/source/finetune.md [hunk 5]
```diff
@@ -63,4 +63,38 @@ to visualise the data,
...
+    3) Inside the newly created efficient-transformers/dataset/custom_dataset.py, define a function named 'get_custom_dataset'. 
+    4) get_custom_dataset() should have following 4 parameters:  dataset_config, tokenizer, split, context_length. This function gets called twice through Qefficient/cloud/finetune.py with the name get_preprocessed_dataset. 
+    5) Inside get_custom_dataset(), dataset needs to prepared for fine tuning. So, the user needs to apply prompt and tokenize the dataset accordingly. Please refer the below template on how to define get_custom_dataset().
+    6) For examples, please refer python files present in efficient-transformers/QEfficient/finetune/dataset. In case of Samsum dataset, get_preprocessed_samsum() of efficient-transformers/QEfficient/finetune/dataset/samsum_dataset.py is called. 
+    7) In efficient-transformers/QEfficient/finetune/configs/dataset_config.py, for custom_dataset class, pass the appropriate value for train_split and test_split according to the dataset keys corresponding to train and test data points.
```
Comment (quic-meetkuma) [inline docs/source/finetune.md:78]: I think this is no longer needed after PR#289. We can directly pass --train_split and --test_split from the CLI.

**Review by quic-meetkuma — CHANGES_REQUESTED** [review]
Good work in listing down the detailed steps for custom dataset, Swati! Please check on the comments. :)

File: docs/source/finetune.md (hunk 1 above)
Comment (quic-swatia) [reply docs/source/finetune.md]: Removed.

File: docs/source/finetune.md (hunk 2 above)
Comment (quic-swatia) [reply docs/source/finetune.md:94]: Added.

File: docs/source/finetune.md (hunk 3 above)
Comment (quic-swatia) [reply docs/source/finetune.md:88]: Added.

File: docs/source/finetune.md (hunk 4 above)
Comment (quic-swatia) [reply docs/source/finetune.md:91]: Added.

File: docs/source/finetune.md (hunk 5 above)
Comment (quic-swatia) [reply docs/source/finetune.md]: Yes, correct! Thanks for pointing this.  After PR 289, user can do this through both ways. Updated step #7 accordingly.

**Review by quic-meetkuma — APPROVED** [review]
Looks good to me.

File: docs/source/finetune.md [hunk 6]
```diff
@@ -63,4 +63,40 @@ to visualise the data,
...
+
+To run fine tuning for any user specific dataset, prepare the dataset using the following steps:
+
+    1) Create a directory named 'dataset' inside efficient-transformers. 
+    2) Inside this directory, create a file named 'custom_dataset.py'. This is different than the custom_dataset.py present at efficient-transformers/QEfficient/finetune/dataset.
+    3) Inside the newly created efficient-transformers/dataset/custom_dataset.py, define a function named 'get_custom_dataset'. 
+    4) get_custom_dataset() should have following 4 parameters:  dataset_config, tokenizer, split, context_length. This function gets called twice through Qefficient/cloud/finetune.py with the name get_preprocessed_dataset. 
+    5) Inside get_custom_dataset(), dataset needs to prepared for fine tuning. So, the user needs to apply prompt and tokenize the dataset accordingly. Please refer the below template on how to define get_custom_dataset().
+    6) For examples, please refer python files present in efficient-transformers/QEfficient/finetune/dataset. In case of Samsum dataset, get_preprocessed_samsum() of efficient-transformers/QEfficient/finetune/dataset/samsum_dataset.py is called.
```
[the remark this answers is not in the document]
Comment (quic-swatia) [reply docs/source/finetune.md:77]: The steps I have mentioned matches with the format of samsum_dataset.py. It doesn't match with alpaca_dataset.py. Hence, I didn't change it.

File: docs/source/finetune.md [hunk 7]
```diff
@@ -63,4 +63,40 @@ to visualise the data,
...
 tensorboard --logdir runs/<file> --bind_all
-```
\ No newline at end of file
+```
+
+## Fine-Tuning on custom dataset
+
+To run fine tuning for any user specific dataset, prepare the dataset using the following steps:
+
+    1) Create a directory named 'dataset' inside efficient-transformers.
```
[the remark this answers is not in the document]
Comment (quic-swatia) [reply docs/source/finetune.md]: Done.

File: docs/source/finetune.md [hunk 8]
```diff
@@ -63,4 +63,40 @@ to visualise the data,
...
+```
+
+## Fine-Tuning on custom dataset
+
+To run fine tuning for any user specific dataset, prepare the dataset using the following steps:
+
+    1) Create a directory named 'dataset' inside efficient-transformers. 
+    2) Inside this directory, create a file named 'custom_dataset.py'. This is different than the custom_dataset.py present at efficient-transformers/QEfficient/finetune/dataset.
+    3) Inside the newly created efficient-transformers/dataset/custom_dataset.py, define a function named 'get_custom_dataset'. 
+    4) get_custom_dataset() should have following 4 parameters:  dataset_config, tokenizer, split, context_length. This function gets called twice through Qefficient/cloud/finetune.py with the name get_preprocessed_dataset.
```
[the remark this answers is not in the document]
Comment (quic-swatia) [reply docs/source/finetune.md]: Done.

**Review by quic-mamta — APPROVED** [review]

File: docs/source/finetune.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (anujgupt-github) [reply docs/source/finetune.md:77]: too verbose. Make it simple pointed steps

File: docs/source/finetune.md [hunk 9]
```diff
@@ -63,4 +63,40 @@ to visualise the data,
...
+    4) get_custom_dataset() should have following 4 parameters:  dataset_config, tokenizer, split, context_length. This function gets called twice through QEfficient/cloud/finetune.py with the name get_preprocessed_dataset. 
+    5) Inside get_custom_dataset(), dataset needs to prepared for fine tuning. So, the user needs to apply prompt and tokenize the dataset accordingly. Please refer the below template on how to define get_custom_dataset().
+    6) For examples, please refer python files present in efficient-transformers/QEfficient/finetune/dataset. In case of Samsum dataset, get_preprocessed_samsum() of efficient-transformers/QEfficient/finetune/dataset/samsum_dataset.py is called. 
+    7) In efficient-transformers/QEfficient/finetune/configs/dataset_config.py, for custom_dataset class, pass the appropriate value for train_split and test_split according to the dataset keys corresponding to train and test data points. As an alternative, these values can be passed as command line arguemnets as well with the finetune command. For example "--train_split train".
```
Comment (anujgupt-github) [inline docs/source/finetune.md:78]: Add hyperlinks to the relative paths annotated in the steps below

File: docs/source/finetune.md [hunk 10]
```diff
@@ -63,4 +63,40 @@ to visualise the data,
 
 ```python
 tensorboard --logdir runs/<file> --bind_all
-```
\ No newline at end of file
+```
+
+## Fine-Tuning on custom dataset
```
Comment (anujgupt-github) [inline docs/source/finetune.md:68]: You should include details on how we use gradient accumulation, how the dataset is shuffled, how activation checkpointing is enabled in separate sections. 
In custom dataset, add a point that if any user wants to use these, refer xyz section
